### ISSUE TITLE
[Proposal] [candle_nn] Redesign Module/ModuleT traits

### ISSUE THREAD

**Issue by nkoppel** [huggingface/candle#1506]
While implementing #1504, I noticed that the the blanket implementation of ModuleT for Modules makes it awkward to create different behavior between training and learning. For a module to use the train parameter of the ModuleT trait, it cannot implement Module, because doing so will cause a "conflicting implementations" error at compile time. I believe that removing the ModuleT trait and adding the `forward_t` method to the Module trait can make this api easier to use.
```rust
trait Module {
    fn forward(&self, xs: &Tensor) -> Result<Tensor>;
    fn forward_t(&self, xs: &Tensor, _train: bool) -> Result<Tensor> {
        self.forward(xs)
    }
}
```
Note that the above implementation is nearly the same to the end user but with more flexibility. Implementations of `forward` only would look exactly the same, while implementations of `forward_t` only would look like the following:
```rust
impl Module for Mlp {
    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
        self.forward_t(xs, false)
    }

    fn forward_t(&self, xs: &Tensor, train: bool) -> Result<Tensor> {
        // ... long, complicated implementation passing 'train' to submodules
    }
}
```
We could even make the implementation of `forward` above the default implementation, though this could lead to users doing `impl Module for Mlp {}` and creating infinite recursion.

These changes would not be difficult to implement, but I understand if they need to wait until a major/minor release.

**Comment by LaurentMazare** [comment]
Actually the original idea behind `Module` and `ModuleT` is a bit different from what you suggest: the key thing is that we should never have an implicit `Module` implementation that would pick one of the two modes used in `ModuleT` as this would be really error prone not knowing if the implicit default is to use train or eval mode (PyTorch would use train by default for example). The opposite is safe though: if a type implements `Module`, it doesn't care about whether it's in train or eval mode so the `forward` implementation can be used for both.

So in order to fix the issue for `BatchNorm` I just made #1513 which removes the `Module` implementation for `BatchNorm` in favor of `ModuleT`, this is a bit annoying as it forces every call to batch-norm to pick its mode but it's a good thing that the api users have in mind what they actually use.

**Comment by nkoppel** [comment]
Okay, I agree that forcing the user to think about training mode is a good thing, and I see that my implementation could lead to confusion due to users calling `forward` when they shouldn't. As for making  `BatchNorm` convenient to use with the `Module` trait, perhaps we could add wrappers so that structs which implement `ModuleT` can easily be used with `Module` like so:
```rust
pub struct Training<M: ModuleT>(M);
pub struct Inference<M: ModuleT>(M);

impl<M: ModuleT> Module for Training<M> { ... }
impl<M: ModuleT> Module for Inference<M> { ... }
```
I also have some concerns about the current implementation of `Sequential`, mainly that it does not support `ModuleT` modules. Considering how simple `Sequential` is, I think it would make sense to simply duplicate some code and make a `SequentialT` struct instead of trying to generalize it somehow.

**Comment by nerdai** [comment]
+1 on the having a variant of `Sequential` that supports `ModuleT`.
